enable optional use of an HTTP proxy

thomasgl-orange opened the pull request:
I need to run the `haproxy_exporter` from an environment which can only reach the target haproxy stats API by going through an HTTP proxy. The way `haproxy_exporter` currently executes its GET requests (with a customized `http.Transport` to allow disabling TLS validation) ignores the proxy settings defined in the traditional environment variables (`$HTTP(S)_PROXY`, `$NO_PROXY`).  This PR fixes that, by adding the proxy from [`http.ProxyFromEnvironment`](https://pkg.go.dev/net/http#ProxyFromEnvironment) to the `http.Transport`. If there is no proxy env vars, or the target haproxy domain is in `$NO_PROXY`, the proxy will be `nil`, and nothing changes in the way requests are executed.

It can be tested, for instance, by running a local [mitmproxy](https://mitmproxy.org/) service, and running `haproxy_exporter` with a `HTTP_PROXY=http://localhost:8080` (or `$HTTPS_PROXY`) env var.

CC @grobie, @roidelapluie

roidelapluie commented:
This is a breaking change and not how we usually do things in the Prometheus ecosystem. I think it would be preferable to add a proxy URL to the URL parameters, or to a specific file (to avoid leaking secrets).

thomasgl-orange commented:
> This is a breaking change 

It is a breaking change for someone who runs `haproxy_exporter` with an `$HTTP(S)_PROXY` var set, but no (proper) `$NO_PROXY` var, and who relies on these proxy settings being ignored. I think it is a very unlikely situation, with straightforward solutions.

> not how we usually do things in the Prometheus ecosystem

This PR simply restores what would otherwise be the default behaviour of `http.Client`, if the `http.Transport` had not been customized to allow disabling TLS check in #86. `$HTTP_PROXY` and friends were indeed properly taken into account before that (just tested with the 0.8.0 binary), so the current state looks more like an accident than a conscious decision (which is fine, I understand that having to go through an HTTP proxy from the exporter is not a common requirement, and that this "regression" was overlooked back then).

> I think it would be preferable to add a proxy URL to the URL parameters, or to a specific file (to avoid leaking secrets).

I disagree, I don't like having every single binary invent its own way to get proxy settings. I like the standardized env-based way golang promotes in the `http` lib.  Just my opinion of course.

====

On a side note, I realize that my PR could be much shorter:
```diff
diff --git a/haproxy_exporter.go b/haproxy_exporter.go
index 54b4cbb..7b44548 100644
--- a/haproxy_exporter.go
+++ b/haproxy_exporter.go
@@ -332,7 +332,10 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
 }
 
 func fetchHTTP(uri string, sslVerify bool, timeout time.Duration) func() (io.ReadCloser, error) {
-	tr := &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: !sslVerify}}
+	tr := &http.Transport{
+		Proxy:           http.ProxyFromEnvironment,
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: !sslVerify},
+	}
 	client := http.Client{
 		Timeout:   timeout,
 		Transport: tr,
```
I didn't want to push-force anything now that you've already viewed the initial version, but I'm happy to do it if you don't mind.

roidelapluie commented:
Wheter or not it is a breaking change is not really subject of question. Someone could have a global proxy set but targetting haproxy which should not be proxied.

In the whole Prometheus ecosystem, we do not set proxy based on environment variables. I would avoid having one exporter behaving differently than others, but I do recognize that users should be able to set proxies. What about a boolean flag --http.proxy-from-env maybe?

grobie reviewed (approved):
Thanks for the review @roidelapluie and the quick turnaround @thomasgl-orange. The proposed solution looks good to me, I'll leave the final review to @roidelapluie.

roidelapluie commented:
Thanks! There are linting issues.

Can we keep it http.proxy-from-env? that way we could reuse it for other exporters that might need this. It also clarifies that it does not concern HAProxy when used with unix sockets.

thomasgl-orange commented:
The CLI flag is now renamed `http.proxy-from-env`.

Regarding linting error, I don't get any with `golangci-lint` latest version (1.45.2), which is what I had installed here (hence no error for me on `make`).  Downgrading to 1.42.0, I've been able to reproduce (that's with go 1.18.1). It's not related to this PR, I get the same on master.
Anyway, I propose bumping  `golangci-lint` to 1.45.2 (Makefile and GitHub actions). I've added a commit for that, so you can see it fixes the GitHub Action. But I can get rid of this commit if you prefer a separate PR.

roidelapluie commented on README.md:
````diff
@@ -58,6 +58,15 @@ you can disable it using the `--no-haproxy.ssl-verify` flag:
 ```
 
+If scraping a remote HAProxy must be done via an HTTP proxy, you can enable reading of the
+standard `$HTTP_PROXY` / `$HTTPS_PROXY` / `$NO_PROXY` environment variables by using the
````
```suggestion
standard [`$http_proxy` / `$https_proxy` / `$no_proxy` environment variables](https://pkg.go.dev/net/http#ProxyFromEnvironment) by using the
```

thomasgl-orange replied:
+1 for the documentation link, and no opinion about uppercase/lowercase variables names (so I will go with your preference).

roidelapluie commented on haproxy_exporter.go:
```diff
@@ -252,7 +252,7 @@ type Exporter struct {
 
 // NewExporter returns an initialized Exporter.
-func NewExporter(uri string, sslVerify bool, selectedServerMetrics map[int]metricInfo, excludedServerStates string, timeout time.Duration, logger log.Logger) (*Exporter, error) {
+func NewExporter(uri string, sslVerify bool, proxyFromEnv bool, selectedServerMetrics map[int]metricInfo, excludedServerStates string, timeout time.Duration, logger log.Logger) (*Exporter, error) {
```
```suggestion
func NewExporter(uri string, sslVerify, proxyFromEnv bool, selectedServerMetrics map[int]metricInfo, excludedServerStates string, timeout time.Duration, logger log.Logger) (*Exporter, error) {
```

roidelapluie commented on haproxy_exporter.go:
```diff
@@ -331,8 +331,11 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
 }
 
-func fetchHTTP(uri string, sslVerify bool, timeout time.Duration) func() (io.ReadCloser, error) {
+func fetchHTTP(uri string, sslVerify bool, proxyFromEnv bool, timeout time.Duration) func() (io.ReadCloser, error) {
```
```suggestion
func fetchHTTP(uri string, sslVerify, proxyFromEnv bool, timeout time.Duration) func() (io.ReadCloser, error) {
```

roidelapluie commented on haproxy_exporter.go:
```diff
@@ -566,6 +569,7 @@ func main() {
 		haProxyServerExcludeStates = kingpin.Flag("haproxy.server-exclude-states", "Comma-separated list of exported server states to exclude. See https://cbonte.github.io/haproxy-dconv/1.8/management.html#9.1, field 17 statuus").Default(excludedServerStates).String()
 		haProxyTimeout             = kingpin.Flag("haproxy.timeout", "Timeout for trying to get stats from HAProxy.").Default("5s").Duration()
 		haProxyPidFile             = kingpin.Flag("haproxy.pid-file", pidFileHelpText).Default("").String()
+		httpProxyFromEnv           = kingpin.Flag("http.proxy-from-env", "Flag that enables using HTTP proxy settings from environment variables ($HTTP_PROXY, $HTTPS_PROXY, $NO_PROXY)").Default("false").Bool()
```
```suggestion
		httpProxyFromEnv           = kingpin.Flag("http.proxy-from-env", "Flag that enables using HTTP proxy settings from environment variables ($http_proxy, $https_proxy, $no_proxy)").Default("false").Bool()
```

roidelapluie reviewed (commented):
Thanks. golint will be updated by our bot, after which you will be able to rebase your PR on top of main.

thomasgl-orange commented:
For what it's worth (just scratching an itch), a `git bisect` on `golangci-lint` from 1.43.0 to [1.44.0](https://github.com/golangci/golangci-lint/releases/tag/v1.44.0) (first version where linting errors disappear when running on `haproxy_exporter`) leads to golangci/golangci-lint#2373 (bump to [gofumpt v0.2.0](https://github.com/mvdan/gofumpt/releases/tag/v0.2.0)) as the commit which fixed these errors.

thomasgl-orange commented:
Thanks for the review and suggestions; I've applied them all, but forgot `Signed-off-by` because I did it in GitHub web UI (todogroup/gh-issues#50), and so DCO not happy. I will rewrite that when rebasing on an updated master (with golangci-lint bump).

roidelapluie commented:
done :)

thomasgl-orange commented:
Thanks, rebased :)

roidelapluie commented:
Thanks!
